### ISSUE TITLE
CP1252 encoded documents cause lsp-mode to send invalid characters, crashing clangd

### ISSUE THREAD

**Issue by tinloaf** [emacs-lsp/lsp-mode#2870]
**Describe the bug**

When I open a codepage 1252 encoded C++ file that contains e.g. the character "…" in a comment, lsp-mode stops working and I get the error message
```
LSP :: Error from the Language Server: trying to get AST for non-added document (Invalid Parameters)
```

**To Reproduce**

Rename the attached file [cp1252.txt](https://github.com/emacs-lsp/lsp-mode/files/6493425/cp1252.txt) to `cp1252.cpp` (GitHub does not allow attaching `.cpp` files directly…), open it and let `lsp-mode` start. The file contains an `…` character encoded in CP1252.

**Expected behavior**
LSP-Mode to start without errors.

**Which Language Server did you use**
clangd 12

**OS**
Ubuntu Focal inside a WSL 1 container.

**Error callstack**
This is the `documentAdded` message sent by lsp-mode to clangd: 

```
[Trace - 01:20:09 PM] Sending notification 'textDocument/didOpen'.
Params: {
  "textDocument": {
    "uri": "file:///home/lba/tmp/cp1252.cpp",
    "languageId": "cpp",
    "version": 0,
    "text": "int main() {\n\t// This is a comment with a strange character: \205\n}\n"
  }
```

Note the `\205`, which indicates that lsp-mode sends the raw byte `0x85` (which is `…` in CP1252) to clangd. Since clangd and lsp-mode (probably?) communicate in UTF-8, this is an invalid byte.

For more information, here is the complete IO log: 
[lsp-log.txt](https://github.com/emacs-lsp/lsp-mode/files/6493450/lsp-log.txt)

**Comment by petr-tik** [comment]
this feels more of a bug report for the server than the lsp-mode client. 

lsp-mode cannot do much when clangd crashes, so I suggest you open an issue on the clangd issue tracker
https://github.com/clangd/clangd/issues

**Comment by yyoncho** [comment]
most likely duplicate of this - https://github.com/emacs-lsp/lsp-mode/issues/2080

**Comment by tinloaf** [comment]
I don't think so. Looking at the [LSP specs](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#headerPart), LSP has a `Content-Type` header field. If that header field is not sent by a client (and as far as I can tell, `lsp-mode` does not set the header field), the default is `application/vscode-jsonrpc; charset=utf-8`.

So basically, `lsp-mode` is (implicitly) telling `clangd` that it's going to send utf-8, and then sends a byte that is not valid `utf-8`. Sure, `clangd` should probably fail graciously, but at this point, I would say that it's `lsp-mode`'s fault.

I see the problem that `lsp-mode` cannot really know what the original encoding of the file was. My Emacs does not seem to know (it displays a little questionmark-in-diamond-symbol instead of `…`). But wouldn't it be an option for `lsp-mode` to filter out any bytes which are not valid utf-8 before sending? That way, it would adhere to the LSP protocol.

**Comment by yyoncho** [comment]
> But wouldn't it be an option for `lsp-mode` to filter out any bytes which are
> not valid utf-8 before sending? That way, it would adhere to the LSP protocol.

I would rather not do conversions like this runtime.



I tested the same with vscode and it seems like vscode is changing the file encoding (and char encoding) right after you open that file. This is what I see after I do `M-x describe-char` after I open the file after that in emacs.

```
             position: 64 of 67 (94%), column: 49
            character: � (displayed as �) (codepoint 65533, #o177775, #xfffd)
              charset: unicode (Unicode (ISO10646))
code point in charset: 0xFFFD
               syntax: w 	which means: word
             category: .:Base
             to input: type "C-x 8 RET fffd" or "C-x 8 RET REPLACEMENT CHARACTER"
          buffer code: #xEF #xBF #xBD
            file code: #xEF #xBF #xBD (encoded by coding system utf-8-unix)
              display: by this font (glyph code)
    xfthb:-SIL -Padauk Book-normal-normal-normal-*-16-*-*-*-*-0-iso10646-1 (#x2D6)

Character code properties: customize what to show
  name: REPLACEMENT CHARACTER
  general-category: So (Symbol, Other)
  decomposition: (65533) ('�')

There are text properties here:
  face                 (tree-sitter-hl-face:comment)
  fontified            t

[back]

```

**Comment by yyoncho** [comment]
> > But wouldn't it be an option for `lsp-mode` to filter out any bytes which are
> > not valid utf-8 before sending? That way, it would adhere to the LSP protocol.
>
> I would rather not do conversions like this runtime.

Also, some servers crash if you send save file notification and they find different content on the disk compared to what we have sent via didChange notifications
